phac-nml/irida-next
# Workflow: Pipeline configuration

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in a pipeline configuration which will register the available pipelines that can be launched within IRIDA Next. The pipeline configuration is stored in `config/pipelines/pipeline.json`. For each pipeline and version a `nextflow_schema` and `schema_input` file are retrieved from the github repo for the pipeline and are stored under `private/pipelines/REPO_ORGANIZATION_AND_NAME/VERSION/`. Along with these files a `status.json` file is generated and stored which has the `etag` header values for both files, which is used to determine if a newer schema file needs to be downloaded when the server is started up.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

![image](https://github.com/phac-nml/irida-next/assets/25466211/f455c4f0-5686-4f21-808b-7fbd54af3364)

**Version 1.0.2**

![image](https://github.com/phac-nml/irida-next/assets/25466211/9f1636aa-0b33-4c07-beb4-959eda07182a)

**Version 1.0.0**

![image](https://github.com/phac-nml/irida-next/assets/25466211/525584bf-1efd-4186-b560-d11b7f8ecf4d)

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Verify that you don't have a `private/pipelines/` directory
2. Start the IRIDA Next server
3. You should now see that there is a `private/pipelines/phac-nml/iridanextexample` directory with sub directories for the 3 versions in the pipelines.json file (1.0.0, 1.0.1, 1.0.2)
4. Verify in each version directory there are 3 files: `nextflow_schema.json`, `schema_input.json`, and `status.json`
5. Shutdown the server
6. To test that an existing file with the same etag as the file on the server is not downloaded again, open `lib/irida/pipelines.rb` and after the line `return if resource_etag_exists(schema_file_url, pipeline_schema_files_path, type)` in the `write_schema_file` method, put a print statement such as `puts "WRITING FILE" and save the file
7. Start the IRIDA Next server
8. Watch the console to see that there are no "WRITING FILE" lines
9. Login as any user
10. Create a project and sample
11. Select the sample from the project `Samples` page and click the rocket icon
12. Verify 3 pipelines are displayed (Versions 1.0.2, 1.0.1, 1.0.0)
13. Click through the different pipelines (Version 1.0.2 and 1.0.0) should have different parameter forms

**Note: In the dev environment the Register Pipelines is called once for each process (web, css, and job)**

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
